src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -108,219 +287,397 @@ public virtual async Task InvokeAsync(HttpContext context, RequestDelegate next)
+    /// Writes a '401 Access denied.' message to the output when the user fails the policy check.
+    /// </summary>
+    protected virtual Task HandleNotAuthorizedPolicyAsync(HttpContext context, RequestDelegate next, AuthorizationResult authorizationResult)
+        => WriteErrorResponseAsync(context, HttpStatusCode.Unauthorized, new AccessDeniedError("schema") { PolicyRequired = Options.AuthorizedPolicy, PolicyAuthorizationResult = authorizationResult });
+
+    /// <summary>
+    /// Writes a '400 JSON body text could not be parsed.' message to the output.
+    /// Return <see langword="false"/> to rethrow the exception or <see langword="true"/>
+    /// if it has been handled.
+    /// </summary>
+    protected virtual async ValueTask<bool> HandleDeserializationErrorAsync(HttpContext context, RequestDelegate next, Exception exception)
     {
-        // nothing to do in this middleware
-        return Task.CompletedTask;
+        await WriteErrorResponseAsync(context, HttpStatusCode.BadRequest, new JsonInvalidError(exception));
+        return true;
     }
 
-    protected virtual Task RequestExecutedAsync(in GraphQLRequestExecutionResult requestExecutionResult)
+    /// <summary>
+    /// Writes a '400 Batched requests are not supported.' message to the output.
+    /// </summary>
+    protected virtual Task HandleBatchedRequestsNotSupportedAsync(HttpContext context, RequestDelegate next)
+        => WriteErrorResponseAsync(context, HttpStatusCode.BadRequest, new BatchedRequestsNotSupportedError());
+
+    /// <summary>
+    /// Writes a '400 Invalid WebSocket sub-protocol.' message to the output.
+    /// </summary>
+    protected virtual Task HandleWebSocketSubProtocolNotSupportedAsync(HttpContext context, RequestDelegate next)
+        => WriteErrorResponseAsync(context, HttpStatusCode.BadRequest, new WebSocketSubProtocolNotSupportedError(context.WebSockets.WebSocketRequestedProtocols));
+
+    /// <summary>
+    /// Writes a '415 Invalid Content-Type header: could not be parsed.' message to the output.
+    /// </summary>
+    protected virtual Task HandleContentTypeCouldNotBeParsedErrorAsync(HttpContext context, RequestDelegate next)
+        => WriteErrorResponseAsync(context, HttpStatusCode.UnsupportedMediaType, new InvalidContentTypeError($"value '{context.Request.ContentType}' could not be parsed."));
+
+    /// <summary>
+    /// Writes a '415 Invalid Content-Type header: non-supported type.' message to the output.
+    /// </summary>
+    protected virtual Task HandleInvalidContentTypeErrorAsync(HttpContext context, RequestDelegate next)
+        => WriteErrorResponseAsync(context, HttpStatusCode.UnsupportedMediaType, new InvalidContentTypeError($"non-supported media type '{context.Request.ContentType}'. Must be '{MEDIATYPE_JSON}', '{MEDIATYPE_GRAPHQL}' or a form body."));
+
+    /// <summary>
+    /// Indicates that an unsupported HTTP method was requested.
+    /// Executes the next delegate in the chain by default.
```

Comment:
🤔 Temporary impl?

Reply:
The uncommented implementation is the correct implementation.  I have some code there commented out so that anyone looking at the code can see what an alternative implementation might be.  But it is set up so that if you set HandleGet to false, then the UI middleware can handle the request (and similar).  Any request not handled by any middleware will return NotFound by ASP.NET Core.

Reply:
Agree. I still read this code and comments later when I migrate my projects.

Reply:
Unfortunately it does not return MethodNotAllowed status code.  I think I already tried some stuff like adding code after the call to `next(context)` but it didn't work properly so in the end I left it as you see now.  I also noted that even if this code was implemented, it would be inaccurate with the ASP.NET Core CORS implementation, as OPTIONS is handled but only if you've installed the CORS middleware and this code does not know whether CORS handling is in use or not.  If you want to try again, you can write a subsequent PR and see how it goes.